README.md
# World Flags

The world-flags gem includes css files for all the main country flags in the world in pixel size 16, 24, 32, 48 and 64.


## Requirements

This gem can be used with Rails 3.1+ using the asset pipeline.


## Configuration

In your asset `application.css` manifest file:

```css
/*
 *= require_self
 *= require_tree .
 *= require flags/basic
 *= require flags/flags16
 *= require flags/flags32
*/
```

The `flags/basic` stylesheet sets up a basic css for use with borders around the flag images (to mark selected language). Use this css as inspiration and customize by overriding styles as needed.


## Use

To render a single country flag in html, you can use:
```html
<i class="flag f16 ar"/>
```
where 'f16' refers to size and 'ar' can be replaced by any country code.

WorldFlags supports flag sizes in 16, 24, 32, 48 and 64 pixels (size).
The countries corresponding to the codes can be found at [ISO 3166-1 alpha-2](http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)

There is also support for semi-transparent flags. This can be used to fade
certain flags while having the selected flag (or hovered over flag) in full
brightness. Simply add or remove the "semi" class for the flag to adjust the
brightness level (fx for selection/mouse over):
```html
<i class="flag f16 ar semi"/>
```

## Advanced use with Rails helpers

You can use built in helper methods to render lists of flags.
Using these helper methods, the flags will be rendered in HTML as:

 ```html
<ul class="f32">
    <li class="flag ar selected" data-cc="ar" data-country_name="Argentina" data-language_name="Spanish" data-locale="ar">&nbsp;</li>
...
</ul>
```

Example using the flags helper:

```haml
= flag_list 16 do
	= flags :ar, :gb, :selected => :gb
```

Alternatively, using the flag helper:

```haml
= flag_list 32 do
	= flag(:ar, 'Argentina')
	= flag(:gb, 'England', :selected => true)
```

Or using the locale mappings for the label...

```haml
= flag_list 32 do
  = flag(:ar)
  = flag(:gb, :selected => true)
```

Or using the #flag_code helper

```haml
	= flag(:ar, 'Argentina') + flag(:gb, 'England', :selected => flag_code(I18n.locale)
```

You can also include the :with_semi => true option in order to have flags not selected displayed with the 'semi' class (semi-bright image)

For use with tooltips or similar js plguins, it can be useful to set the title attribute:

```haml
= flag_list 32 do
	= flags :ar, :br, :gb, :title => true
```

The flag_title will render the following list item:

```html
<li class="flag ar" lang="ar" title="Argentina">&nbsp;</li>
```

Note: The '&amp;nbsp;' is needed in order for the background (flag icon) to have something to be displayed against.

To get content rendered,

```haml
= flags :ar, :br, :gb, :content => true
```

The :title and :content can also be set to a string which is then displayed

```haml
= flag :ar, 'Argentina', :title => 'Argentina country', :content => 'Argh!'
```

or using the locale mapping...

```haml
= flag :ar, :title => 'Argentina country'
```

Note: There is also a #flag_selected? helper, which is (and/or can be) used to determine if the flag to be drawn should have the "selected" class set)

### Configuration

To disable use of country- and language name data attributes in output:

```ruby
WorldFlags.country_name_disable!
WorldFlags.language_name_disable!
```

### Customizing output

You can customize the output by the flag view helper methods:

```ruby
WorldFlags.flag_list_tag = :div
WorldFlags.flag_tag = :span
WorldFlags.flag_text = ''
```

To do more customization, look at the `world_flags/helper/view/util.rb` file.

### Using localization

You can specify whether to look up labels for the flags for either language or country and for which locale to look up the labels (see Configuring localization)

Use danish (da) country labels

```haml
= flag_list 32 do
	= flags :ar, :br, :gb, :country => :da
```

Use language labels for current locale.
Note: Also use border class (see CSS below)

```haml
= flag_list 32, :class => 'border' do
	= flags :ar, :br, :gb, :language => I18n.locale
```

In the /config folder of this gem/engine there is a json file with all the english _ISO-3166-2_ code translations ready for use. You can make similar locale files for other locales/languages. You can use either of the following:

* [countries_and_languages](https://github.com/grosser/countries_and_languages)
* [i18n data](https://github.com/grosser/i18n_data)
* [i18n gen](https://github.com/kristianmandrup/i18n-gen)

To generate i18n data for your particular locale(s).

## Configuration and usage tips

See the [wiki](https://github.com/kristianmandrup/world-flags/wiki/)

* [Configuration](https://github.com/kristianmandrup/world-flags/wiki/Configuration)
* [Handling flag selection](https://github.com/kristianmandrup/world-flags/wiki/Handling-flag-selection)

## Geo IP

Include the 'geoip' gem if you want to detect locale based on browser IP location.

[GeoIP country lite](http://www.maxmind.com/app/geolitecountry)
[GeoIP](http://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz)

Extract and put latest `GeoIP.dat` file in `db/GeoIP.dat` of your Rails app. Alternatively set location via `WorldFlags.geo_ip_db_path = path`. 

The _world-flags_ engine comes pre-packaged with a recent version of `GeoIP.dat` in the `db` folder of the engine itself.

### GeoIP generator

Copy WorldFlags GeoIP.dat to Rails app '/db/GeoIP.dat'

TODO: Should optimally fetch the one from maxmind.com

`rails g world_flags:geoip`

### Localhosts filtering

Also set `WorldFlags.localhost_list` if you have localhosts other than the default `127.0.0.1`.

You can override the default ip passed to the `Geo module by overriding the `browser_ip` method in your controller. Here the default implementation is shown:

### Fine control of the IP used

```ruby
def browser_ip
  request.remote_ip
end
```

Code extract of the locale source logic for browser IP country code detection:

```ruby
when :ip
  country_code_from_ip browser_ip
```

## CSS

The `basic.css` file in the `vendor/assets/stylesheets/flags` folder of this repo defines the basic styling of flags. Override these styles to fit your needs.

```ruby
.flags {
  list-style-type: none;
  padding: 0;
  margin-left: 0;  
}

.flags .l {
  float: left;
}

.border .flag {
  border: 1px solid;  
}

.border .selected {
  border: 1px solid black;
}

.f16.flag {
  width: 16px;
  height: 16px;
}

.f24.flag {
  width: 24px;
  height: 24px;
}

 /* and so on ... */
```

## Sprites

The sprites contains all the main country flags in the world.

You can use the [sprite generator](http://spritegen.website-performance.org/) to generate sprites for other icons sets and follow the pattern for this gem.

You can also see the [pictos-icons](https://github.com/kristianmandrup/pictos-icons)  
or [social_icons](https://github.com/kristianmandrup/social_icons) gems for other examples using this model.

Also check out the [world-flag-packs](https://github.com/kristianmandrup/world-flag-packs) repo, that contains multiple flag packs ready for use with `world-flags.

## Customizing Flag sprite size

*Resize images:*

Use ImageMagick to resize, using 32px version as base

`convert flags32.png -resize 75% flags24.png`

`convert flags32_semi.png -resize 75% flags24_semi.png`

*Generate adsjusted css files:*

The CSS files must have positioning adjusted to fit the new sprites:
Use the resizing tool in `lib/world-flags/tools/resize_css.rb`
Edit the last part where ResizeCss is initialized

`resizer = ResizeCss.new 32, 24`

The last argument (here 24) is the flag size used for calculating new positioning in the css files generated.

The ResizeCSS initializer takes many additional options, including :auto-exec to auto convert the sprite using ImageMagick.

`resizer = ResizeCss.new 32, 24, auto-exec: true`

See the code for more details on how to adjust to fit your needs.

Run the resizer

`$ ruby resize_css.rb`

TODO: refactor resize css tool into rake task or similar that takes arguments!

## TODO for version 1.0

Suggestions welcome ;)

## Enjoy

Copyright (2012) Kristian Mandrup



CHANGELOG
## 1.0.0

With this release using the asset pipeline is required.

- Source moved to https://github.com/world-flags/world-flags
- Allow rendering single flags
- Add runtime dependency on sass-rails
- Support for Rails 4

## 0.2.0

Lots of new features, including:

- locale helper for use in Controllers
- localization/translations of countries and languages
- auto-selection of flag class, for flag matching current locale
- manually set which flag is to be selected
- and much more...




lib/world_flags/helper/view.rb
require 'world_flags/helper/view/util'

module WorldFlags
	module Helper
		module View
			def self.flag_sizes
				[16, 24, 32, 48, 64]
			end

			def flags_list size = 24, options = {}, &block
				raise "Missing block" unless block_given?
				unless WorldFlags::Helper::View.flag_sizes.include?(size.to_i)
					raise "Supported sizes are only #{WorldFlags::Helper::View.flag_sizes}" 
				end
				content = capture(&block)
				xclass = options[:class] ? " #{options[:class]}" : ''
				content_tag WorldFlags.flag_list_tag, content, :class => "f#{size} flags#{xclass}"
			end
			alias_method :flag_list, :flags_list

			# http://en.wikipedia.org/wiki/ISO_639-1_language_matrix

			# should look up translation for each code
			def flags *args
				options = args.extract_options!
				args.flatten!
				args.inject("") do |res, elem|
					case elem
					when String, Symbol
						code = elem
						name = WorldFlags.label(code, options)
					else
						raise ArgumentError, "Bad argument: #{args}, must be Array"
					end				
					res << flag(code, name, options)
				end.html_safe
			end

			def flags_title *args
				options = args.extract_options!
				flags args, options.merge(:title => true)
			end

			def flag code, *args
				options = args.extract_options!
				name = args.first.kind_of?(String) ? args.first : WorldFlags.label(code, options)				

				label = WorldFlags::Helper::View::Util.label_for name, options
				title = WorldFlags::Helper::View::Util.title_for name, options

				content_tag WorldFlags.flag_tag, label.html_safe, WorldFlags::Helper::View::Util.flag_options(code, title, name, options)
			end

			def flag_title code, *args
				options = args.extract_options!
				name = args.first.kind_of?(String) ? args.first : WorldFlags.label(code, options)				

				flag code, name, options.merge(:title => true)
			end

			def use_flags size = 24
				stylesheet_link_tag "flags/flags#{size}"
			end
		end
	end
end


lib/world_flags/helper/view/util.rb
module WorldFlags
	module Helper
		module View
			module Util
				def self.label_for name, options = {}
					label = case options[:content]
					when true 
						name
					when String
						options[:content]
					else
						WorldFlags.flag_text
					end
				end				

				def self.title_for name, options = {}
					case options[:title]
					when true
						name
					when String
						options[:title]
					else
						nil
					end				
				end

				def self.flag_options code, title, name, options = {}
					locale = I18n.locale
					extra_options = title ? {:title => title } : {}			
					selected = flag_selected?(code, options) ? ' selected' : ''

					# add semi class if not selected
					semi = (selected.blank? ? ' semi' : '') if options[:with_semi]
					xclass = options[:class] ? " #{options[:class]}" : ''

					if WorldFlags.country_name_enabled?
						country_name = WorldFlags.country(code, locale)						
					end
					country_option = country_name ? {:'data-country_name' => country_name} : {}

					if WorldFlags.language_name_enabled?
						language_name = WorldFlags.language(code, locale)
					end
					language_option = language_name ? {:'data-language_name' => language_name} : {}

					flag_locale = WorldFlags.locale(code)

					{:class => "flag #{code}#{selected}#{semi}#{xclass}", :'data-cc' => code, :'data-locale' => flag_locale}.merge(options[:html] || {}).merge(extra_options).merge(country_option).merge(language_option)
				end

				def self.flag_selected? code, options = {}
					code = code.to_sym
					sel = options[:selected] || options[code]

					auto_sel = flag_code(I18n.locale).to_sym if WorldFlags.auto_select?
					match_locales = [sel, auto_sel]

					selected ||= match_locales.any?{|e| e == code }
				end

				# define more mappings by setting WorldFlags.locale_flag_map to a Hash map
				# http://en.wikipedia.org/wiki/ISO_639-1_language_matrix
				def self.flag_code code
					WorldFlags.flag_code code
				end				
			end
		end
	end
end


lib/generators/world_flags/publicize_generator.rb
module WorldFlags
  module Generators
    class PublicizeGenerator < ::Rails::Generators::Base
      desc "Publicizes world flags assets (if not using Rails asset pipeline)"

      class_option :stylesheets_dir,  :type => :string, 
                                      :desc => 'The stylesheets folder of your app'

      class_option :images_dir,       :type => :string, 
                                      :desc => 'The images folder of your app'

      class_option :javascripts_dir,  :type => :string, 
                                      :desc => 'The javascripts folder of your app'

      class_option :public_dir,       :type => :string, 
                                      :desc => 'The public root folder of your app'

      def main_flow
        copy_dir asset_src_dir('stylesheets'), stylesheets_dir
        copy_dir asset_src_dir('images'), images_dir
        copy_dir asset_src_dir('javascripts'), images_dir
      end    

      protected

      def copy_dir src, target
        say "Copying: #{src} -> #{target}"        
        copy src, target
      end

      def asset_src_dir folder
        File.expand_path "../../../app/assets/#{folder}"
      end

      def javascripts_dir
        options[:stylesheets_dir] || asset_folder('javascripts')
      end

      def stylesheets_dir
        options[:stylesheets_dir] || asset_folder('stylesheets')
      end

      def images_dir
        options[:images_dir] || asset_folder('images')
      end

      def asset_folder folder = nil
        File.join(public_dir, folder)
      end


      def public_dir
        Rails.root.join(options[:public_dir] || 'public')
      end
    end
  end
end



lib/world_flags/tools/resize_css.rb
# use ImageMagick to Resize, using 32px version as base
# convert flags32.png -resize 75% flags24.png
# convert flags32_semi.png -resize 75% flags24_semi.png

require 'open3'

# TODO: Needs some heavy refactoring and cleaning up!

class ResizeCss

  attr_reader :base_size, :new_size, :stylesheet_folder, :stylesheet_name, :sprite_name
  attr_reader :template_lang, :size_prefix, :options

  def initialize base_size, new_size, options
    @base_size  = base_size
    @new_size   = new_size
    @options    = options

    @stylesheet_folder = options[:stylesheet_folder]  if options[:stylesheet_folder]
    @stylesheet_name   = options[:stylesheet_name]    if options[:stylesheet_name]
    @sprite_name       = options[:sprite_name]        if options[:sprite_name]
    @template_lang     = options[:template_lang]      if options[:template_lang]
    @size_prefix       = options[:size_prefix]        if options[:size_prefix]
    
    resize = "convert #{sprite_name}#{base_size}.png -resize #{factor * 100}%"
    new_file_name = "#{sprite_name}#{new_size}"

    command_resize = "#{resize} #{new_file_name}.png"
    command_resize_light = "#{resize} #{lighten} #{new_file_name}_light.png"
    command_resize_dark  = "#{resize} #{darken} #{new_file_name}_dark.png"
    command_resize_gray  = "#{resize} #{grayscale} #{new_file_name}_gray.png"

    if auto_exec?
      Open3.popen3("pwd") do |stdin, stdout, e, t|
        base_dir = stdout
        Open3.popen3 "cd #{images_dir}"

        Open3.popen3 command_resize
        Open3.popen3 command_resize_light if lighten?
        Open3.popen3 command_resize_dark if darken?
        Open3.popen3 command_resize_gray if gray? 

        Open3.popen3 "cd #{base_dir}"
      end
    else
      puts "Use ImageMagick to resize #{sprite_name} sprites:"  
      puts command_resize
      puts command_resize_light if lighten?
      puts command_resize_dark if darken?
      puts command_resize_gray if gray? 
    end
  end

  def lighten?
    options[:light]
  end

  def darken?
    options[:dark]
  end

  def gray?
    options[:gray]
  end

  def lighten
    "-level 0%,100%,2.0"
  end

  def darken
    "-level 0%,100%,0.5"
  end

  def grayscale
    "-colorspace Gray"
  end

  def auto_exec?
    options[:auto_exec]
  end

  def size_prefix
    @size_prefix ||= 'f'
  end

  def app_assets_path
    "../../../app/assets/"
  end

  def images_dir
    @images_dir ||= File.join(app_assets_path, 'images')
  end

  def stylesheets_dir
    @stylesheets_dir ||= File.join(app_assets_path, 'stylesheets')
  end

  def stylesheets_folder
    @stylesheets_folder ||= "flags"
  end

  def stylesheet_full_path
    @stylesheet_full_path ||= File.join(stylesheets_dir, stylesheets_folder)
  end

  def stylesheet_name
    @stylesheet_name ||= 'flags'
  end

  def sprite_name
    @sprite_name ||= 'flags'
  end

  def template_lang
    @template_lang ||= 'erb'
  end

  def template_ext
    !blank?(template_lang) ? ".#{template_lang}" : ''
  end

  def blank? txt
    !txt || txt.empty?
  end

  def execute options = {:ext => 'css', :semi => false}
    semi = options[:semi] ? '-semi': ''
    ext = options[:ext] ? options[:ext] : 'css'
    src_file_path = File.expand_path "#{stylesheet_full_path}/#{stylesheet_name}#{base_size}#{semi}.#{ext}#{template_ext}", File.dirname(__FILE__)
    target_file_path = File.expand_path "#{stylesheet_full_path}/#{stylesheet_name}#{new_size}#{semi}.#{ext}#{template_ext}", File.dirname(__FILE__)

    puts "src: #{src_file_path} -> #{target_file_path}"

    lines = File.open(src_file_path).readlines

    img_exp = "#{sprite_name}-#{base_size}"
    replace_img = "#{sprite_name}-#{new_size}"

    # puts "img_exp: #{img_exp}"
    # puts "replace_img: #{replace_img}"

    repositioned_css = lines.map do |line|
      pos_match = line.match /.+position:\s-?(\d+)(px)?\s-?(\d+)(px)?/
      # puts "pos_match: #{line} - #{pos_match.inspect}"
      img_match = line.match /#{img_exp}/
      new_line = line.sub(/#{img_exp}/, replace_img) if img_match
      new_line = replace_pos pos_match, line if pos_match
      # puts "line: #{line} -> #{new_line}"
      new_line ||= line
      new_line
    end.join("")

    puts "Write new CSS:"
    puts repositioned_css
    
    File.open(target_file_path, 'w') do |f|
      f.puts repositioned_css.gsub ".#{size_prefix}#{base_size}", ".#{size_prefix}#{new_size}"
    end
  end

  def factor
    @factor ||= new_size / base_size.to_f
  end

  def replace_pos pos_match, line
    pos_x = pos_match[1].to_i        
    new_pos_x = (pos_x * factor).to_i.to_s

    pos_y = pos_match[3].to_i        
    new_pos_y = (pos_y * factor).to_i.to_s
    
    # puts "new pos: #{new_pos_x}, #{new_pos_y}"
    
    rest = line[6..-1]
    rest = rest.sub /#{pos_x}/, new_pos_x.to_s
    rest = rest.sub /#{pos_y}/, new_pos_y.to_s

    rest = rest.gsub /#{base_size}px;/, "#{new_size}px;"

    # first = line[0..10].sub(".f#{base_size}", ".f#{newsize}") 
    line[0..5] + rest
  end
end
# options
# - :stylesheet_folder, :stylesheet_name, :sprite_name, :template_lang, :size_prefix
# - :dark, :light, :gray
# - :auto-exec

# change last argument to fit sprite size of flags
# first arg is the flag size (css files) that is used as base
resizer = ResizeCss.new 24, 48


resizer.execute 
resizer.execute :ext => 'css.scss'

# resizer.execute :semi => true
# resizer.execute :ext => 'css.scss', :semi => true
